### PR TITLE
Introduce `ControllerRevisionUpgrade` to `instancetype.kubevirt.io`

### LINKED ISSUE #9909: Implement a conversion strategy for `instancetype.kubevirt.io/v1alpha{1,2}` objects stored in `ControllerRevisions` to `instancetype.kubevirt.io/v1beta1`
/area instancetype

**What happened**:

$subject, existing `instancetype.kubevirt.io/v1alpha{1,2}` objects stored in `ControllerRevisions` are not upgraded to `v1beta1` making the removal of support for these older versions impossible.

**What you expected to happen**:

**How to reproduce it (as minimally and precisely as possible)**:
Steps to reproduce the behavior.

**Additional context**:

**Environment**:
- KubeVirt version (use `virtctl version`): N/A
- Kubernetes version (use `kubectl version`): N/A
- VM or VMI specifications: N/A
- Cloud provider or hardware configuration: N/A
- OS (e.g. from /etc/os-release): N/A
- Kernel (e.g. `uname -a`): N/A
- Install tools: N/A
- Others: N/A

### PR DESCRIPTION
**Pull request by lyarwood**
/area instancetype
/cc @0xFelix 

**What this PR does / why we need it**:

This PR introduces a new `ControllerRevisionUpgrade` CRD to `instancetype.kubevirt.io` and the associated `ControllerRevisionUpgradeController`  controller to `virt-controller` to handle the upgrade of `ControllerRevisions` containing old  versions of `instancetype.kubevirt.io` objects.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #9909

**Special notes for your reviewer**:

instancetype: Add ControllerRevision migration proposal
https://github.com/kubevirt/community/pull/226

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [x] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [x] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
A new `ControllerRevisionUpgrade` CRD has been introduced to `instancetype.kubevirt.io/v1beta1` to orchastrate the upgrade of `ControllerRevisions` containing older versions of `instancetype.kubevirt.io` objects
```

### CODE REVIEW

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (0xFelix) [inline pkg/virt-controller/watch/instancetype/upgrade.go:125]: Did I understand correctly that Queue can contain a key multiple times? So when it is reenqueued this does not remove the reenqueued key?

File: pkg/virt-controller/watch/instancetype/upgrade_test.go
Comment (0xFelix) [inline pkg/virt-controller/watch/instancetype/upgrade_test.go:196]: Will it retry indefinitely?

**Review by 0xFelix — COMMENTED** [review]
Does a CR always have only one owner? Can it have multiple, especially when we dedup CRs in the future? What happens if the owner goes away, will the revision be deleted too? Is that problematic for DR?

Can VirtualMachine{Instancetype,Preference}SpecRevision be in a CR too? Do we have upgrade code for that?

**Comment by lyarwood** [comment]
Thanks for the excellent review as always @0xFelix!

> Does a CR always have only one owner?

At present yes.

> Can it have multiple, especially when we dedup CRs in the future? 

Yes in the future we will have multiple owners per CR and this upgrade process will need to handle that, copying all owner references over but I don't want to handle that in this PR.

> What happens if the owner goes away, will the revision be deleted too? Is that problematic for DR?

Yes, at present if the owner is deleted the CR is also removed. I'm not sure how that would be an issue for DR as the CR is owned by the VM not the VMI?

> Can VirtualMachine{Instancetype,Preference}SpecRevision be in a CR too? Do we have upgrade code for that?

Yes see https://github.com/kubevirt/kubevirt/pull/10458 that's a follow up to this at the moment given the size of this PR and the need to refactor the compat code to remove `isPreference` finally.

Looking at the rest of your comments now, thanks again!

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade.go:125]: No I think it's a single key but `AddRateLimited` / `Add` [1] marks the key as dirty (requires processing) that is the handled by `Done` [2]  with the key added back into the queue again. ewwww.

[1]  https://github.com/kubevirt/kubevirt/blob/1b9232fada18fb9dfaac67a73a2ea880f188e5e3/vendor/k8s.io/client-go/util/workqueue/queue.go#L119-L139
[2] https://github.com/kubevirt/kubevirt/blob/1b9232fada18fb9dfaac67a73a2ea880f188e5e3/vendor/k8s.io/client-go/util/workqueue/queue.go#L177-L193

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade.go:177]: This code will ensure the key is forgotten immediately but yeah in the failure case we will keep retrying.

This is copied from the other `virt-controller` controllers but I don't think it fits our usecase so I'm happy replacing this with fail and forget.

File: pkg/virt-controller/watch/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade_test.go:196]: Yeah as above, I don't mind reworking this to fail and forget.

File: pkg/virt-controller/watch/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade_test.go:243]: Yeah as above, I'll change this behaviour.

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (lyarwood) [inline staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:589]: I've somehow forgot to wire this up, I'll do that now.

File: tests/subresource_api_test.go
Comment (lyarwood) [reply tests/subresource_api_test.go:436]: https://github.com/kubevirt/kubevirt/pull/10456/commits/b6c87f80179c42e12ff2d66009ed15a3bcb0e5df

> These helpers will be moved into a new libinstancetype package shortly
> so instead of importing the instancetype tests here just inline the
> creation of these resources for now.

File: tests/instancetype/upgrade.go
Comment (lyarwood) [reply tests/instancetype/upgrade.go:165]: No, I'll add coverage for unlabelled CRs now.

File: pkg/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/instancetype/upgrade_test.go:70]: Nope

File: pkg/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/instancetype/upgrade_test.go:216]: At the time I wrote this I think I was using `GenerateRevisionNamePatch` in `upgrade.go` but I've since switched to `GenerateTestReplacePatch` so we should be able to handle that here.

File: pkg/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/instancetype/upgrade_test.go:247]: I feel like there was a reason `Raw` had to be `nil` as well but I can't recall what it was at the moment, I'll finish the other cleanups and take another look at this.

File: pkg/virt-controller/watch/application.go
Comment (lyarwood) [reply pkg/virt-controller/watch/application.go:38]: Ah sorry no `instancetype` is already taken by the import below.

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade.go:58]: We need to replace it in the tests with a mock queue.

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (lyarwood) [reply staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:589]: Done

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/virt-controller/watch/instancetype/upgrade.go:177]: Removed.

File: pkg/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/instancetype/upgrade.go:205]: Done.

File: tests/instancetype/upgrade.go
Comment (0xFelix) [inline tests/instancetype/upgrade.go:199]: This will not update the CR on the API which the UpgradeController is going to look at?

File: tests/instancetype/upgrade.go
Comment (0xFelix) [inline tests/instancetype/upgrade.go:325]: Same as above

File: tests/instancetype/upgrade.go
Comment (0xFelix) [inline tests/instancetype/upgrade.go:415]: Same as above.

File: tests/instancetype/upgrade.go
Comment (0xFelix) [inline tests/instancetype/upgrade.go:531]: Same as above.

File: pkg/instancetype/upgrade_test.go
Comment (akrejcir) [inline pkg/instancetype/upgrade_test.go:472]: My comments in the previous table apply here too.

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (akrejcir) [inline pkg/virt-controller/watch/instancetype/upgrade.go]: This function was removed, so this commit will not compile.

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (akrejcir) [inline staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:574]: Can you squash this change with the commit that defines these constants?

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (akrejcir) [inline pkg/virt-controller/watch/instancetype/upgrade.go:244]: Can you pass `string` parameter instead of `error`? The error is only used as message passed to the condition.

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (akrejcir) [inline pkg/virt-controller/watch/instancetype/upgrade.go:256]: nit: Use `%w` to wrap error.

**Review by akrejcir — COMMENTED** [review]
Functionality looks good. I have mostly code style comments.

What component creates the `ControllerRevisionUpgrade` CRD? Will it be in a future PR?

In my opinion, the way this PR is split to commits makes it harder to review:
- Sometimes one commit adds code and another commit modifies it, instead of adding the correct code in the first commit.
- Code is added in one commit that is not used anywhere, and then another commit uses it.

Can you add more text to commit messages?

The `nit:` comments are only a suggestion. You can ignore them, if you want.

File: pkg/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/instancetype/upgrade.go:76]: @0xFelix added this and after some initial thought today I don't think we need this.

It's working around an issue in our functional tests where I was using the current `CreateControllerRevision` implementation to create legacy `ControllerRevisions`.

This used the current naming convention and would cause naming conflicts in the functional tests when we tried to run an upgrade on unlabelled `ControllerRevisions` containing `v1beta1` objects. However these `ControllerRevisions` wouldn't use the new naming convention introduced by this PR and thus the combination IMHO is invalid.

I've updated the functional tests to produce legacy `ControllerRevisions` when needed now using the older naming convention and removing labels as these also were never present prior to `v1beta1`.

File: tests/instancetype/upgrade.go
Comment (lyarwood) [reply tests/instancetype/upgrade.go:34]: Yeah agreed but I'm not sure what to do here tbh, we do have coverage for everything in unit tests already but I still feel the functional coverage adds significant value here.

**Review by lyarwood — COMMENTED** [review]
Thanks for the excellent review @akrejcir!

> Functionality looks good. I have mostly code style comments.
> 
> What component creates the `ControllerRevisionUpgrade` CRD? Will it be in a future PR?

https://github.com/kubevirt/community/pull/226 goes into some limited detail on this but a second `ControllerRevisionUpgradeTrigger` CRD will automate the creation of these individual `ControllerRevisionUpgrade` requests. I'll refresh the design document in the morning as naming has changed since I wrote that and many of the TODOs are now resolved IMHO.

> In my opinion, the way this PR is split to commits makes it harder to review:
> 
> * Sometimes one commit adds code and another commit modifies it, instead of adding the correct code in the first commit.
> * Code is added in one commit that is not used anywhere, and then another commit uses it. Then I cannot review the commits separately.

ACK apologies, that was just a pre-PTO rush on my end. I've squashed and reorganised things so they are hopefully easier to follow now.

> Can you add more text to commit messages?

I'll handle this in the next respin, again apologies.

> The `nit:` comments are only a suggestion. You can ignore them, if you want.

I think I've mostly handled these thanks!

File: pkg/util/status/status.go
Comment (akrejcir) [reply pkg/util/status/status.go:392]: Maybe a maintainer can answer why is this pattern useful, and if it is ok to remove it?

File: pkg/virt-controller/watch/instancetype/upgrade_test.go
Comment (akrejcir) [reply pkg/virt-controller/watch/instancetype/upgrade_test.go:112]: Right, sorry I missed that the default `UpgradeFn` is not a noop, but sets a new name.
It is enough.

File: tests/instancetype/upgrade.go
Comment (akrejcir) [reply tests/instancetype/upgrade.go:37]: Ok, we can keep them.

File: pkg/instancetype/upgrade.go
Comment (0xFelix) [inline pkg/instancetype/upgrade.go:49]: This should have `Upgrader` as return type.

File: pkg/instancetype/upgrade.go
Comment (0xFelix) [inline pkg/instancetype/upgrade.go:81]: Can this condition still be `false`? If so the `Create` before would have failed already.

File: pkg/instancetype/upgrade.go
Comment (akrejcir) [reply pkg/instancetype/upgrade.go:154]: Ok, I missed that.
I think using the same package name caused a package dependency error. Maybe it's solved now.

File: pkg/instancetype/upgrade_test.go
Comment (0xFelix) [inline pkg/instancetype/upgrade_test.go:483]: You removed the deletion of CRs with the same name, this will not pass?

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (0xFelix) [inline pkg/virt-controller/watch/instancetype/upgrade.go:222]: nit: Rename to `upgradeStatusUnknown`?

File: pkg/virt-controller/watch/instancetype/upgrade.go
Comment (0xFelix) [inline pkg/virt-controller/watch/instancetype/upgrade.go:226]: nit: Rename to `upgradeStatusSuccessful`?

File: pkg/instancetype/upgrade.go
Comment (lyarwood) [reply pkg/instancetype/upgrade.go:81]: I don't believe so unless someone manually removes the labels from a `ControllerRevision` created after this PR lands and we then try to upgrade the `ControllerRevision` while it contains an object of the latest version.

I don't think we need to defend against that situation here as it complicates the code, the upgrade would just fail in that case and could be easily fixed.

File: pkg/instancetype/upgrade_test.go
Comment (lyarwood) [reply pkg/instancetype/upgrade_test.go:483]: The removal of the original CR still happens when we switch things out to a new CR.

**Review by 0xFelix — APPROVED** [review]
Thanks!

/lgtm

File: staging/src/kubevirt.io/api/instancetype/v1beta1/types.go
Comment (akrejcir) [inline staging/src/kubevirt.io/api/instancetype/v1beta1/types.go:587]: I forgot to comment on this in my last review.
Usually, in other objects, `Spec` and `Status` fields are not pointers. Can you remove the pointers?

**Comment by xpivarc** [comment]
Left a comment on the proposal.

**Comment by lyarwood** [comment]
/hold

While we look at doing this on `resync`.

**Comment by lyarwood** [comment]
@xpivarc @0xFelix @akrejcir alternative draft implementation upgrading on resync posted https://github.com/kubevirt/kubevirt/pull/10545

**Pull request closed without merging**
